Mathlib/MeasureTheory/Integral/SobolevInequality.lean

```diff
@@ -0,0 +1,661 @@
+    Fintype.one_lt_card_iff_nontrivial.mp (by exact_mod_cast hp.one_lt)
+  have : (1:ℝ) ≤ ↑#ι - 1 := by
+    have hι : (2:ℝ) ≤ #ι := by exact_mod_cast hp.one_lt
+    linarith
+  calc ∫⁻ x, (‖u x‖₊ : ℝ≥0∞) ^ p
+      = ∫⁻ x, ((‖u x‖₊ : ℝ≥0∞) ^ (1 / (#ι - 1 : ℝ))) ^ (#ι : ℝ) := by
+        -- a little algebraic manipulation of the exponent
+        congr! 2 with x
+        rw [← ENNReal.rpow_mul, hp.conj_eq]
+        field_simp
+    _ = ∫⁻ x, ∏ _i : ι, (‖u x‖₊ : ℝ≥0∞) ^ (1 / (#ι - 1 : ℝ)) := by
+        -- express the left-hand integrand as a product of identical factors
+        congr! 2 with x
+        simp_rw [prod_const, card_univ]
+        norm_cast
+    _ ≤ ∫⁻ x, ∏ i, (∫⁻ xᵢ, ‖fderiv ℝ u (update x i xᵢ)‖₊) ^ ((1 : ℝ) / (#ι - 1 : ℝ)) := ?_
+    _ ≤ (∫⁻ x, ‖fderiv ℝ u x‖₊) ^ p :=
+        -- apply the grid-lines lemma
+        lintegral_prod_lintegral_pow_le _ hp (by fun_prop)
+  -- we estimate |u x| using the fundamental theorem of calculus.
+  gcongr with x i
+  calc (‖u x‖₊ : ℝ≥0∞)
+      = (‖∫ xᵢ in Iic (x i), deriv (u ∘ update x i) xᵢ‖₊ : ℝ≥0∞) := by
+        -- apply the half-infinite fundamental theorem of calculus
+        have h3u : ContDiff ℝ 1 (u ∘ update x i) := hu.comp (by convert contDiff_update 1 x i)
+        have h4u : HasCompactSupport (u ∘ update x i) :=
+          h2u.comp_closedEmbedding (closedEmbedding_update x i)
+        simp [HasCompactSupport.integral_Iic_deriv_eq h3u h4u (x i)]
+    _ ≤ ∫⁻ xᵢ in Iic (x i), ‖deriv (u ∘ update x i) xᵢ‖₊ :=
+        ennnorm_integral_le_lintegral_ennnorm _ -- apply the triangle inequality
+    _ ≤ ∫⁻ xᵢ, (‖fderiv ℝ u (update x i xᵢ)‖₊ : ℝ≥0∞) := ?_
+  gcongr with y; swap; exact Measure.restrict_le_self
+  -- bound the derivative which appears
+  calc ‖deriv (u ∘ update x i) y‖₊ = ‖fderiv ℝ u (update x i y) (deriv (update x i) y)‖₊ := by
+        rw [fderiv.comp_deriv _ (hu.differentiable le_rfl).differentiableAt
+          (hasDerivAt_update x i y).differentiableAt]
+    _ ≤ ‖fderiv ℝ u (update x i y)‖₊ * ‖deriv (update x i) y‖₊ :=
+        ContinuousLinearMap.le_opNNNorm ..
+    _ ≤ ‖fderiv ℝ u (update x i y)‖₊ := by simp [deriv_update, Pi.nnnorm_single]
+
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] [MeasurableSpace E] [BorelSpace E]
+  [FiniteDimensional ℝ E] (μ : Measure E) [IsAddHaarMeasure μ]
+
+open FiniteDimensional
+
+/-- The **Gagliardo-Nirenberg-Sobolev inequality**.  Let `u` be a continuously differentiable
+compactly-supported function `u` on a normed space `E` of finite dimension `n ≥ 2`, equipped
+with Haar measure. There exists a constant `C` depending only on `E`, such that the Lebesgue
+integral of the pointwise expression `|u x| ^ (n / (n - 1))` is bounded above by `C` times the
+`n / (n - 1)`-th power of the Lebesgue integral of the Fréchet derivative of `u`. -/
+theorem lintegral_pow_le_pow_lintegral_fderiv {p : ℝ} (hp : Real.IsConjExponent (finrank ℝ E) p) :
+    ∃ C : ℝ≥0, ∀ {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] [CompleteSpace F]
```

Comment:
I have a small problem with existential statements like that, that the constant `C` might a priori depend on the universe of `F`. So you can not really write that "it only depend on E and mu" (well, it's true, but not what the statement says). I wonder if it's not clearer to define a constant `gagliardoNirenbergSobolevConstant E \mu := ...` and use that in the statements, to lift this ambiguity (with a shorter notation to be used in the current file, to make things readable). It would also have the advantage to avoid the ugly forall quantifier in the statement.

Same thing for the other statements below.

What do you think?

Reply:
I am in favor, since I think that will be a lot nicer to work with in practice. I'm also doing that in the formalization of Carleson's theorem. I wasn't sure if that style was appropriate for Mathlib.

Reply:
Since multiple statements in this file have slightly different constants, I'll probably use something like `lintegralPowLePowLintegralFderivConst`.